server/lib/device/device.downsamplingStates.js

```diff
@@ -0,0 +1,120 @@
+
+    deviceFeatures.forEach((feature) => {
+      // date interval
+      let beginDate;
+      if (feature.last_downsampling) {
+        beginDate = new Date(feature.last_downsampling);
+      } else {
+        beginDate = new Date(feature.created_at);
+      }
+      beginDate.setHours(0, 0, 0, 0);
+
+      const endDate = new Date();
+      endDate.setHours(0, 0, 0, 0);
+
+      logger.debug(`Date interval of state to treat: ${beginDate} to ${endDate}`);
+
+      // Build query for get yesterday feature states of feature
+      const statesQueryParams = {
+        raw: true,
+        where: {
+          device_feature_id: { [Op.eq]: feature.id },
+          created_at: { [Op.between]: [beginDate, endDate] },
+        },
+        order: [['created_at', 'ASC']],
+      };
+
+      // Delete possible featue state light in conflict
+      const queryInterface = db.sequelize.getQueryInterface();
+      queryInterface.bulkDelete('t_device_feature_state_light', {
+        device_feature_id: { [Op.eq]: feature.id },
+        created_at: { [Op.between]: [beginDate, endDate] },
+      });
+
+      // Get feature state and downsample it to save in feature state light
+      db.DeviceFeatureState.findAll(statesQueryParams, { transaction: t }).then((data) => {
+        logger.debug(data.length);
+
+        if (data.length > 0) {
+          // Split state by day to treat downsampling day by day
+          const mapOfFeatureState = new Map();
+          data.forEach((state) => {
+            const currentDayStateDate = new Date(state.created_at);
+            currentDayStateDate.setHours(0, 0, 0, 0);
+
+            if (!mapOfFeatureState.get(currentDayStateDate.getTime())) {
+              mapOfFeatureState.set(currentDayStateDate.getTime(), []);
+            }
+            mapOfFeatureState.get(currentDayStateDate.getTime()).push(state);
+          });
+
+          mapOfFeatureState.forEach((value) => {
+            if (value.length > MAX_STATE_BY_DAY) {
+              const newFeatureStateArray = [];
+              value.forEach((state) => {
+                newFeatureStateArray.push({
+                  x: new Date(state.created_at).getTime(),
+                  y: state.value,
+                });
+              });
+
+              const smoothFeatureStates = LTTB(newFeatureStateArray, MAX_STATE_BY_DAY);
+              // Rebuild feature state list
+              const featureStatesToSave = [];
+              smoothFeatureStates.forEach((state) => {
+                featureStatesToSave.push({
+                  device_feature_id: feature.id,
+                  created_at: state.x,
+                  value: state.y,
+                });
+              });
+
+              db.DeviceFeatureStateLight.bulkCreate(featureStatesToSave);
+            } else {
+              // In this case all feature state was seed in feature state light
+              db.DeviceFeatureStateLight.bulkCreate(value);
```

Comment:
Be careful with asynchronous function, if you don't put await before, it "lost" and errors will be thrown at the global level. 

Re-read your whole PR, it's full of this.

Reply:
in this example those call is in foreach without async call , I need to use something like Promise.each instead ?

Reply:
You need to at least write:

```
await db.DeviceFeatureStateLight.bulkCreate(value);
```

Reply:
I add await